


<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html id="htmlId">
<head>
  <title>Coverage Report :: CancellableThreads</title>
  <style type="text/css">
    @import "../../.css/coverage.css";
  </style>
</head>

<body>
<div class="header"></div>

<div class="content">
<div class="breadCrumbs">
    [ <a href="../../index.html">all classes</a> ]
    [ <a href="../index.html">org.elasticsearch.common.util</a> ]
</div>

<h1>Coverage Summary for Class: CancellableThreads (org.elasticsearch.common.util)</h1>

<table class="coverageStats">

<tr>
  <th class="name">Class</th>
<th class="coverageStat 
">
  Method, %
</th>
<th class="coverageStat 
">
  Line, %
</th>
</tr>
<tr>
  <td class="name">CancellableThreads</td>
<td class="coverageStat">
  <span class="percent">
    90%
  </span>
  <span class="absValue">
    (9/ 10)
  </span>
</td>
<td class="coverageStat">
  <span class="percent">
    55.2%
  </span>
  <span class="absValue">
    (37/ 67)
  </span>
</td>
</tr>
  <tr>
    <td class="name">CancellableThreads$ExecutionCancelledException</td>
<td class="coverageStat">
  <span class="percent">
    0%
  </span>
  <span class="absValue">
    (0/ 2)
  </span>
</td>
<td class="coverageStat">
  <span class="percent">
    0%
  </span>
  <span class="absValue">
    (0/ 4)
  </span>
</td>
  </tr>
  <tr>
    <td class="name">CancellableThreads$Interruptible</td>
  </tr>
  <tr>
    <td class="name">CancellableThreads$IOInterruptible</td>
  </tr>
<tr>
  <td class="name"><strong>total</strong></td>
<td class="coverageStat">
  <span class="percent">
    75%
  </span>
  <span class="absValue">
    (9/ 12)
  </span>
</td>
<td class="coverageStat">
  <span class="percent">
    52.1%
  </span>
  <span class="absValue">
    (37/ 71)
  </span>
</td>
</tr>
</table>

<br/>
<br/>


<div class="sourceCode"><i>1</i>&nbsp;/*
<i>2</i>&nbsp; * Licensed to Elasticsearch under one or more contributor
<i>3</i>&nbsp; * license agreements. See the NOTICE file distributed with
<i>4</i>&nbsp; * this work for additional information regarding copyright
<i>5</i>&nbsp; * ownership. Elasticsearch licenses this file to you under
<i>6</i>&nbsp; * the Apache License, Version 2.0 (the &quot;License&quot;); you may
<i>7</i>&nbsp; * not use this file except in compliance with the License.
<i>8</i>&nbsp; * You may obtain a copy of the License at
<i>9</i>&nbsp; *
<i>10</i>&nbsp; *    http://www.apache.org/licenses/LICENSE-2.0
<i>11</i>&nbsp; *
<i>12</i>&nbsp; * Unless required by applicable law or agreed to in writing,
<i>13</i>&nbsp; * software distributed under the License is distributed on an
<i>14</i>&nbsp; * &quot;AS IS&quot; BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
<i>15</i>&nbsp; * KIND, either express or implied.  See the License for the
<i>16</i>&nbsp; * specific language governing permissions and limitations
<i>17</i>&nbsp; * under the License.
<i>18</i>&nbsp; */
<i>19</i>&nbsp;package org.elasticsearch.common.util;
<i>20</i>&nbsp;
<i>21</i>&nbsp;import org.apache.lucene.util.SetOnce;
<i>22</i>&nbsp;import org.apache.lucene.util.ThreadInterruptedException;
<i>23</i>&nbsp;import org.elasticsearch.ElasticsearchException;
<i>24</i>&nbsp;import org.elasticsearch.common.Nullable;
<i>25</i>&nbsp;import org.elasticsearch.common.io.stream.StreamInput;
<i>26</i>&nbsp;
<i>27</i>&nbsp;import java.io.IOException;
<i>28</i>&nbsp;import java.util.HashSet;
<i>29</i>&nbsp;import java.util.Set;
<i>30</i>&nbsp;
<i>31</i>&nbsp;/**
<i>32</i>&nbsp; * A utility class for multi threaded operation that needs to be cancellable via interrupts. Every cancellable operation should be
<i>33</i>&nbsp; * executed via {@link #execute(Interruptible)}, which will capture the executing thread and make sure it is interrupted in the case
<i>34</i>&nbsp; * of cancellation.
<i>35</i>&nbsp; *
<i>36</i>&nbsp; * Cancellation policy: This class does not support external interruption via &lt;code&gt;Thread#interrupt()&lt;/code&gt;. Always use #cancel() instead.
<i>37</i>&nbsp; */
<b class="fc"><i>38</i>&nbsp;public class CancellableThreads {</b>
<b class="fc"><i>39</i>&nbsp;    private final Set&lt;Thread&gt; threads = new HashSet&lt;&gt;();</b>
<i>40</i>&nbsp;    // needs to be volatile as it is also read outside of synchronized blocks.
<b class="fc"><i>41</i>&nbsp;    private volatile boolean cancelled = false;</b>
<b class="fc"><i>42</i>&nbsp;    private final SetOnce&lt;OnCancel&gt; onCancel = new SetOnce&lt;&gt;();</b>
<i>43</i>&nbsp;    private String reason;
<i>44</i>&nbsp;
<i>45</i>&nbsp;    public synchronized boolean isCancelled() {
<b class="fc"><i>46</i>&nbsp;        return cancelled;</b>
<i>47</i>&nbsp;    }
<i>48</i>&nbsp;
<i>49</i>&nbsp;    public void checkForCancel() {
<b class="fc"><i>50</i>&nbsp;        checkForCancel(null);</b>
<b class="fc"><i>51</i>&nbsp;    }</b>
<i>52</i>&nbsp;
<i>53</i>&nbsp;    private void checkForCancel(Exception beforeCancelException) {
<b class="fc"><i>54</i>&nbsp;        if (isCancelled()) {</b>
<i>55</i>&nbsp;            final String reason;
<i>56</i>&nbsp;            final OnCancel onCancel;
<b class="nc"><i>57</i>&nbsp;            synchronized (this) {</b>
<b class="nc"><i>58</i>&nbsp;                reason = this.reason;</b>
<b class="nc"><i>59</i>&nbsp;                onCancel = this.onCancel.get();</b>
<b class="nc"><i>60</i>&nbsp;            }</b>
<b class="nc"><i>61</i>&nbsp;            if (onCancel != null) {</b>
<b class="nc"><i>62</i>&nbsp;                onCancel.onCancel(reason, beforeCancelException);</b>
<i>63</i>&nbsp;            }
<i>64</i>&nbsp;            // fallback to the default exception
<b class="nc"><i>65</i>&nbsp;            final RuntimeException cancelExp = new ExecutionCancelledException(&quot;operation was cancelled reason [&quot; + reason + &quot;]&quot;);</b>
<b class="nc"><i>66</i>&nbsp;            if (beforeCancelException != null) {</b>
<b class="nc"><i>67</i>&nbsp;                cancelExp.addSuppressed(beforeCancelException);</b>
<i>68</i>&nbsp;            }
<b class="nc"><i>69</i>&nbsp;            throw cancelExp;</b>
<i>70</i>&nbsp;        }
<b class="fc"><i>71</i>&nbsp;    }</b>
<i>72</i>&nbsp;
<i>73</i>&nbsp;    private synchronized boolean add() {
<b class="fc"><i>74</i>&nbsp;        checkForCancel();</b>
<b class="fc"><i>75</i>&nbsp;        threads.add(Thread.currentThread());</b>
<i>76</i>&nbsp;        // capture and clean the interrupted thread before we start, so we can identify
<i>77</i>&nbsp;        // our own interrupt. we do so under lock so we know we don&#39;t clear our own.
<b class="fc"><i>78</i>&nbsp;        return Thread.interrupted();</b>
<i>79</i>&nbsp;    }
<i>80</i>&nbsp;
<i>81</i>&nbsp;    /**
<i>82</i>&nbsp;     * run the Interruptible, capturing the executing thread. Concurrent calls to {@link #cancel(String)} will interrupt this thread
<i>83</i>&nbsp;     * causing the call to prematurely return.
<i>84</i>&nbsp;     *
<i>85</i>&nbsp;     * @param interruptible code to run
<i>86</i>&nbsp;     */
<i>87</i>&nbsp;    public void execute(Interruptible interruptible) {
<i>88</i>&nbsp;        try {
<b class="fc"><i>89</i>&nbsp;            executeIO(interruptible);</b>
<b class="nc"><i>90</i>&nbsp;        } catch (IOException e) {</b>
<b class="nc"><i>91</i>&nbsp;            assert false : &quot;the passed interruptible can not result in an IOException&quot;;</b>
<b class="nc"><i>92</i>&nbsp;            throw new RuntimeException(&quot;unexpected IO exception&quot;, e);</b>
<b class="fc"><i>93</i>&nbsp;        }</b>
<b class="fc"><i>94</i>&nbsp;    }</b>
<i>95</i>&nbsp;    /**
<i>96</i>&nbsp;     * run the Interruptible, capturing the executing thread. Concurrent calls to {@link #cancel(String)} will interrupt this thread
<i>97</i>&nbsp;     * causing the call to prematurely return.
<i>98</i>&nbsp;     *
<i>99</i>&nbsp;     * @param interruptible code to run
<i>100</i>&nbsp;     */
<i>101</i>&nbsp;    public void executeIO(IOInterruptible interruptible) throws IOException {
<b class="fc"><i>102</i>&nbsp;        boolean wasInterrupted = add();</b>
<b class="fc"><i>103</i>&nbsp;        boolean cancelledByExternalInterrupt = false;</b>
<b class="fc"><i>104</i>&nbsp;        RuntimeException runtimeException = null;</b>
<b class="fc"><i>105</i>&nbsp;        IOException ioException = null;</b>
<i>106</i>&nbsp;
<i>107</i>&nbsp;        try {
<b class="fc"><i>108</i>&nbsp;            interruptible.run();</b>
<b class="nc"><i>109</i>&nbsp;        } catch (InterruptedException | ThreadInterruptedException e) {</b>
<i>110</i>&nbsp;            // ignore, this interrupt has been triggered by us in #cancel()...
<b class="nc"><i>111</i>&nbsp;            assert cancelled : &quot;Interruption via Thread#interrupt() is unsupported. Use CancellableThreads#cancel() instead&quot;;</b>
<i>112</i>&nbsp;            // we can only reach here if assertions are disabled. If we reach this code and cancelled is false, this means that we&#39;ve
<i>113</i>&nbsp;            // been interrupted externally (which we don&#39;t support).
<b class="nc"><i>114</i>&nbsp;            cancelledByExternalInterrupt = !cancelled;</b>
<b class="nc"><i>115</i>&nbsp;        } catch (RuntimeException t) {</b>
<b class="nc"><i>116</i>&nbsp;            runtimeException = t;</b>
<b class="nc"><i>117</i>&nbsp;        } catch (IOException e) {</b>
<b class="nc"><i>118</i>&nbsp;            ioException = e;</b>
<i>119</i>&nbsp;        } finally {
<b class="fc"><i>120</i>&nbsp;            remove();</b>
<b class="fc"><i>121</i>&nbsp;        }</b>
<i>122</i>&nbsp;        // we are now out of threads collection so we can&#39;t be interrupted any more by this class
<i>123</i>&nbsp;        // restore old flag and see if we need to fail
<b class="fc"><i>124</i>&nbsp;        if (wasInterrupted) {</b>
<b class="nc"><i>125</i>&nbsp;            Thread.currentThread().interrupt();</b>
<i>126</i>&nbsp;        } else {
<i>127</i>&nbsp;            // clear the flag interrupted flag as we are checking for failure..
<b class="fc"><i>128</i>&nbsp;            Thread.interrupted();</b>
<i>129</i>&nbsp;        }
<b class="fc"><i>130</i>&nbsp;        checkForCancel(ioException != null ? ioException : runtimeException);</b>
<b class="fc"><i>131</i>&nbsp;        if (ioException != null) {</b>
<i>132</i>&nbsp;            // if we&#39;re not canceling, we throw the original exception
<b class="nc"><i>133</i>&nbsp;            throw ioException;</b>
<i>134</i>&nbsp;        }
<b class="fc"><i>135</i>&nbsp;        if (runtimeException != null) {</b>
<i>136</i>&nbsp;            // if we&#39;re not canceling, we throw the original exception
<b class="nc"><i>137</i>&nbsp;            throw runtimeException;</b>
<i>138</i>&nbsp;        }
<b class="fc"><i>139</i>&nbsp;        if (cancelledByExternalInterrupt) {</b>
<i>140</i>&nbsp;            // restore interrupt flag to at least adhere to expected behavior
<b class="nc"><i>141</i>&nbsp;            Thread.currentThread().interrupt();</b>
<b class="nc"><i>142</i>&nbsp;            throw new RuntimeException(&quot;Interruption via Thread#interrupt() is unsupported. Use CancellableThreads#cancel() instead&quot;);</b>
<i>143</i>&nbsp;        }
<i>144</i>&nbsp;
<b class="fc"><i>145</i>&nbsp;    }</b>
<i>146</i>&nbsp;
<i>147</i>&nbsp;
<i>148</i>&nbsp;    private synchronized void remove() {
<b class="fc"><i>149</i>&nbsp;        threads.remove(Thread.currentThread());</b>
<b class="fc"><i>150</i>&nbsp;    }</b>
<i>151</i>&nbsp;
<i>152</i>&nbsp;    /** cancel all current running operations. Future calls to {@link #checkForCancel()} will be failed with the given reason */
<i>153</i>&nbsp;    public synchronized void cancel(String reason) {
<b class="fc"><i>154</i>&nbsp;        if (cancelled) {</b>
<i>155</i>&nbsp;            // we were already cancelled, make sure we don&#39;t interrupt threads twice
<i>156</i>&nbsp;            // this is important in order to make sure that we don&#39;t mark
<i>157</i>&nbsp;            // Thread.interrupted without handling it
<b class="nc"><i>158</i>&nbsp;            return;</b>
<i>159</i>&nbsp;        }
<b class="fc"><i>160</i>&nbsp;        cancelled = true;</b>
<b class="fc"><i>161</i>&nbsp;        this.reason = reason;</b>
<b class="fc"><i>162</i>&nbsp;        for (Thread thread : threads) {</b>
<b class="nc"><i>163</i>&nbsp;            thread.interrupt();</b>
<b class="nc"><i>164</i>&nbsp;        }</b>
<b class="fc"><i>165</i>&nbsp;        threads.clear();</b>
<b class="fc"><i>166</i>&nbsp;    }</b>
<i>167</i>&nbsp;
<i>168</i>&nbsp;
<i>169</i>&nbsp;    public interface Interruptible extends IOInterruptible {
<i>170</i>&nbsp;        void run() throws InterruptedException;
<i>171</i>&nbsp;    }
<i>172</i>&nbsp;
<i>173</i>&nbsp;    public interface IOInterruptible {
<i>174</i>&nbsp;        void run() throws IOException, InterruptedException;
<i>175</i>&nbsp;    }
<i>176</i>&nbsp;
<i>177</i>&nbsp;    public static class ExecutionCancelledException extends ElasticsearchException {
<i>178</i>&nbsp;
<i>179</i>&nbsp;        public ExecutionCancelledException(String msg) {
<b class="nc"><i>180</i>&nbsp;            super(msg);</b>
<b class="nc"><i>181</i>&nbsp;        }</b>
<i>182</i>&nbsp;
<i>183</i>&nbsp;        public ExecutionCancelledException(StreamInput in) throws IOException {
<b class="nc"><i>184</i>&nbsp;            super(in);</b>
<b class="nc"><i>185</i>&nbsp;        }</b>
<i>186</i>&nbsp;    }
<i>187</i>&nbsp;
<i>188</i>&nbsp;    /**
<i>189</i>&nbsp;     * Registers a callback that will be invoked when some running operations are cancelled or {@link #checkForCancel()} is called.
<i>190</i>&nbsp;     */
<i>191</i>&nbsp;    public synchronized void setOnCancel(OnCancel onCancel) {
<b class="nc"><i>192</i>&nbsp;        this.onCancel.set(onCancel);</b>
<b class="nc"><i>193</i>&nbsp;    }</b>
<i>194</i>&nbsp;
<i>195</i>&nbsp;    @FunctionalInterface
<i>196</i>&nbsp;    public interface OnCancel {
<i>197</i>&nbsp;        /**
<i>198</i>&nbsp;         * Called when some running operations are cancelled or {@link #checkForCancel()} is explicitly called.
<i>199</i>&nbsp;         * If this method throws an exception, cancelling tasks will fail with that exception; otherwise they
<i>200</i>&nbsp;         * will fail with the default exception {@link ExecutionCancelledException}.
<i>201</i>&nbsp;         *
<i>202</i>&nbsp;         * @param reason                the reason of the cancellation
<i>203</i>&nbsp;         * @param beforeCancelException any error that was encountered during the execution before the operations were cancelled.
<i>204</i>&nbsp;         * @see #checkForCancel()
<i>205</i>&nbsp;         * @see #setOnCancel(OnCancel)
<i>206</i>&nbsp;         */
<i>207</i>&nbsp;        void onCancel(String reason, @Nullable Exception beforeCancelException);
<i>208</i>&nbsp;    }
<i>209</i>&nbsp;}
</div>
</div>

<div class="footer">
    
    <div style="float:right;">generated on 2020-02-09 18:46</div>
</div>
</body>
</html>
